Quirks for Halium compatibility

fredldotme opened the pull request:
This branch allows to override the behavior of telepathy-ofono via Android's getprop mechanism.
The Sony Xperia X for example requires further configuration to force the output and input/sources and sinks.
This is done using a tweak similar to: https://github.com/fredldotme/device-sony-loire/commit/17b2e5f005def89c48b809b4725487d29b2bf5af
This enables calls both using the earpiece & headset+mic.

fredldotme committed:
telepathy-ofono: Multiple changes

- Quirks for select hardware:
  - Force default sink to "sink.primary" via t-o.quirk.forcesinkprimary=1
  - Force default source to "source.primary" via t-o.quirk.forcesourceprimary=1
- H2W detection:
  - Read h2w switch file for whether to include wired headsets/-phones into in-/output decision
- Always set default sink and source, even if they match the defaults.

fredldotme committed:
QPulseAudioEngine: Improve quirk meaning

The quirk is required on ports where the name has to match a specific in/output name.

Flohack74 commented on qpulseaudioengine.cpp:
```diff
@@ -312,29 +351,43 @@ void QPulseAudioEngineWorker::sinkInfoCallback(const pa_sink_info *info)
     if ((m_audiomode & AudioModeBluetooth) && (modes.contains(AudioModeBluetooth))) {
         preferred = bluetooth_sco;
         audiomodetoset = AudioModeBluetooth;
+        qDebug("Prefer AudioModeBluetooth");
```
I would remove that logging from production code as in general logfiles are just noise on the flash memory and will wear it off ;) - and also log rotation is set so shortly that you probably do not see them after a few days anyways.

fredldotme replied:
The service is dbus-activated, not started by upstart, meaning it's not showing up in typical log files. The output only shows when run manually.

UniversalSuperBox replied:
I must disagree. The wear on the flash is really quite small. For a much worse example of this, do `tail -f /var/log/syslog`, then turn your display on and off.

fredldotme replied:
My mistake, dbus logs are stored in `~/.cache/upstart/dbus.log`.

fredldotme commented:
Checklist of devices this has been tested on without issues:

* [x] Meizu MX4
* [x] Oneplus One
* [ ] Fairphone 2
* [ ] Nexus 5
* [ ] BQ E4.5
* [ ] BQ E5
* [ ] Meizu Pro 5
* [ ] Nexus 4
* [ ] Moto G 2014

NeoTheThird commented:
Any instructions for QA?

fredldotme commented:
> Any instructions for QA?

`sudo ubports-qa install xenial_-_haliumfixup`

EDIT: Additionally the call tests should be done with both headset plugged in and without a headset.

fredldotme commented:
I just noticed that initiating calls with the headset plugged in doesn't work yet because AppArmor denies access to the switch state file.
What's needed to fix this is include telepathy-mission-control-5 in our packaging and add an AppArmor rule for the state file.
~~I will import the repo and add the AppArmor rule shortly, all within another xenial_-_haliumfixup branch.~~
EDIT: Fixed by shipping a local override in telepathy-ofono itself.

fredldotme committed:
telepathy-ofono: Ship AppArmor profile override for h2w detection

Due to the AppArmor profile set in telepathy-mission-control-5 the h2w file
is not allowed to be accessed by default.
Ship a custom override in the telepathy-ofono package to allow access.

fredldotme commented:
Testing can continue now, specifically the headset detection while initiating the call (plugging in the headset before dialing).

UniversalSuperBox commented on CMakeLists.txt:
```diff
@@ -113,12 +114,13 @@ add_dependencies(${TELEPATHY_OFONO} schema_update qrc_update)
 
 enable_testing()
```
Might it also be worth it to add regression tests for these changes?

fredldotme replied:
There are not QPulseAudioEngine tests yet. What would be needed, more than just headset detection?

UniversalSuperBox replied:
Basically anything that does anything should be tested to ensure it does what it reports itself to do... But if there are no tests on this module at all I suspect it was determined to be too difficult or its logic was too trivial. Does it seem to call out to other sources to do anything, or do you think it's relatively self-contained?

fredldotme replied:
QPulseAudioEngine communicates with PulseAudio which might be hard to test thoroughly.

UniversalSuperBox replied:
Okay... *hangs head in no-test shame*

fredldotme committed:
apparmor: Allow property_service socket & install in the right destination

fredldotme committed:
apparmor: Allow relative path to property_socket

telepathy-ofono runs from cwd /, so allow access to the relative path to the properties socket.

fredldotme committed:
QPulseAudioEngine: Remove h2w detection

h2w detection is now properly handled by pulseaudio-modules-droid-24 (extcon).

dobey commented on qpulseaudioengine.cpp:
```diff
@@ -99,6 +100,26 @@ static void subscribeCallback(pa_context *context, pa_subscription_event_type_t
+static char quirk_primary_sink_name[PROP_VALUE_MAX];
+static char* quirk_sinkprimary_name(int& len)
+{
+    if ((len = property_get("t-o.quirk.forcesink", quirk_primary_sink_name, NULL)) > 0) {
```
Could we not instead use gsettings for this, so that it's still usable without hybris?

fredldotme replied:
As this is a quirk for select devices I fail to see how to set gsettings properties from within the Halium port/adaptation.

dobey replied:
Sure, but we also still use telepathy-ofono on devices which do not have hybris. In the former days of Canonical images, the "custom" tarball included gsettings overrides, and is read-only in `/custom` path on the device. GSettings keys can also be "locked" so that generally the key could not be changed on those devices, as well. Unfortunately I think _pulseaudio_ doesn't use gsettings/dconf, so we can't take advantage of overrides and schema lock, without adding keys in our own schema for telepathy-ofono.

fredldotme replied:
Would those override files be required to change when we change a gsettings schema of telepathy-ofono? Just curious.
Also I think the Android properties are a leaner way of handling such quirks. Shouldn't telepathy-ofono work on non-hybris devices even with linking to libandroid-properties?

dobey replied:
The override files would only need to be changed if the schema was changed in some incompatible way. They aren't copies of schemas, but INI style files which specify what keys to override, and with what values, sort of like:
```
[org.compiz.gnomecompat]
main-menu-key=""
run-key=""
```
Even if telepathy-ofono would work without hybris while linking to libandroid-properties, it's something extra we have to include, which I think we can't use anyway. So, IMO, in general it is better to implement these things in some more generic way, so that we can bring the hybris and non-hybris builds even closer together, without just shipping useless things on the non-hybris builds.

mariogrip replied:
How about using https://github.com/ubports/deviceinfo for this? will gsetting work in the case? will it not needed to be compiled?

dobey replied:
We could use `deviceinfo` for this, sure. It would need to be rebuilt every time a change is added, so that new config ends up on devices, and it looks like it needs to be rebuilt for xenial already anyway. It wouldn't use GSettings in this case though, as it's a separate thing.

mariogrip replied:
Doing that will be faily simple,

just 
```
std::shared_ptr<DeviceInfo> info = std::make_shared<DeviceInfo>();

if (info->contains("AudioForceSink"){
  m_nametoset = info->get("AudioForceSink", "").c_str();
}

{same for source}

```


then put  in a folder with the [devicename].conf and place it in /etc/device-info
```
[device]
AudioForceSink=thesink
AudioForceSource=thesrc
```

mariogrip replied:
> We could use `deviceinfo` for this, sure. It would need to be rebuilt every time a change is added, so that new config ends up on devices, and it looks like it needs to be rebuilt for xenial already anyway. It wouldn't use GSettings in this case though, as it's a separate thing.

deviceinfo wont need to be recompiled to change a config? It needs to be included to xenial yes, but that should be fairly easy to do now that it's to a point where its ready.

fredldotme replied:
I've noticed that with failing getprop (due to running telepathy-ofono with the old AppArmor profile pre-reboot, just restarting the service) in-call-audio still works.
As it seems the only change we require on the Xperia X is setting the source/sink even if it matches the default.
We could idealy get rid of the quirk altogether as no other device has been reported to have working audio yet using it.

fredldotme replied:
Claims made before proper testing, seems the quirks are still needed. I'll take a look at deviceinfo. Should the ini file be provided as a halium override? @mariogrip

mariogrip replied:
> Claims made before proper testing, seems the quirks are still needed. I'll take a look at deviceinfo. Should the ini file be provided as a halium override? @mariogrip

If this is a device specific quirk then it will need to be a "device" override. for example if "hammerhead" needed this quirk i would add 
```
[device]
AudioForceSink=thesink
AudioForceSource=thesrc
```
to a file named /etc/device-info/hammerhead.conf

but if it's global to halium then it can be a halium override. I also plan to add another level of override, halium-version override.

mariogrip replied:
In short there is 4 levels of overrides:

device - priority 1
driver (halium or linux) - priority 2
type (desktop, phone, tablet) - priority 3
default - priority 4

fredldotme replied:
I've thought about it further and I personally think that we should proceed with the android property mechanism for the time being. The camera quirks by peat-psuwit for example also use the same mechanism and we shouldn't require porters to add configuration changes all over the place using different mechanisms. We're basically bike-shedding atm.

mariogrip reviewed (approved):
Using android pops are fine for now until device-info gets more mature, once it gets more mature and tested we can move all the things over, but for now LGTM :+1:

mariogrip commented:
@UniversalSuperBox, is it time to get into current devel cycle?

fredldotme committed:
QPulseAudioEngine: use ro prefix for quirks

iambumblehead commented:
@fredldotme @Flohack74 @NeoTheThird @mariogrip @dobey 

I'm lurking here and looking forward to trying out Ubuntu Touch on the Xperia X

As an Xperia X owner, I can confirm that [this video expresses our feelings towards this PR.][0]

[0]: https://www.youtube.com/watch?v=tYzMYcUty6s

Flohack74 commented on qpulseaudioengine.cpp:
```diff
@@ -312,29 +335,43 @@ void QPulseAudioEngineWorker::sinkInfoCallback(const pa_sink_info *info)
     if ((m_audiomode & AudioModeBluetooth) && (modes.contains(AudioModeBluetooth))) {
         preferred = bluetooth_sco;
         audiomodetoset = AudioModeBluetooth;
+        qDebug("Prefer AudioModeBluetooth");
```
Will the logging be supressed in production builds? Otherwise I suggest removing it.

fredldotme replied:
I really need this info for when I'm helping porters debug their stuff and they might not know how to build a new package with debug-level logging enabled.

UniversalSuperBox commented:
Note for QA: It's unlikely, but this could cause issues with audio routing during calls. Try to send or recieve calls using a variety of loudspeaker, earpiece, and headset to test these changes.

ziggutas commented:
BQ E4.5 on rc (2019-W41/2)
Sending receiving calls on loudspeaker and earpiece all ok. I don't have a headset to test.
